### PR TITLE
Add a nccl-test sample workload

### PR DESCRIPTION
**Pull request by Chris113113**
Adds a docker image and helm chart for deploying nccltest.

Support for:
* Multi-node nccl-test via mpirun
* TCPx config values
* Support for any test present in `nccl-test`
* GKE

### CODE REVIEW

File: sample_workloads/nccltest/docker/root_scripts/gen_hostfiles.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+for nr in "${NRANKS_FACTORS[@]}";
+do
+  rm -f "hostfiles${NHOSTS}/hostfile${nr}"
+  touch "hostfiles${NHOSTS}/hostfile${nr}"
+  for h in "$@";
+  do
+    echo "$h port=222 slots=${nr}" >> "hostfiles${NHOSTS}/hostfile${nr}"
+  done
+done
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/docker/root_scripts/gen_hostfiles.sh:24]: Nit: EOF linebreak.

File: sample_workloads/nccltest/docker/root_scripts/init_ssh.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+while true; do
+  host=$1
+  if [[ -z $host ]]; then
+    break
+  fi
+  ssh -o StrictHostKeyChecking=no -p "${PORT}" "$host" \
+    echo "Hello from ${host}"
+  shift
+done
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/docker/root_scripts/init_ssh.sh:13]: Nit: EOF linebreak.

File: sample_workloads/nccltest/docker/root_scripts/tune_net.sh [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  set_irq_range eth4 96 8
+}
+
+a3_bind_irqs
+
+dump_irq_binding eth0
+dump_irq_binding eth1
+dump_irq_binding eth2
+dump_irq_binding eth3
+dump_irq_binding eth4
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/docker/root_scripts/tune_net.sh:63]: Nit: EOF linebreak.

File: sample_workloads/nccltest/docker/root_scripts/tune_net.sh [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+    # this is GVE's RX irq. See gve_rx_idx_to_ntfy().
+    echo "$core" > /proc/irq/$(("$irq" + "$num_q"))/smp_affinity_list
+
+    idx=$(( (idx + 1) % num_cores ))
+  done
+}
+
+a3_bind_irqs() {
+  set_irq_range eth0 32 4
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/docker/root_scripts/tune_net.sh:50]: This is COS specific (network interface naming and vCPU indexing), maybe mention this somewhere in the docs?

File: sample_workloads/nccltest/docker/scripts/container_entry.sh [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+# Job parameters.
+: "${JOB_TIMESTAMP:?Must set JOB_TIMESTAMP}"
+: "${JOB_NAME:?Must set JOB_NAME}"
+: "${MASTER_ADDR:?Must set MASTER_ADDR}"
+: "${NNODES:?Must set NNODES}"
+: "${NODE_RANK:?Must set NODE_RANK}"
+
+# GCS bucket to be used.
+: "${GCS_BUCKET:?Must set GCS_BUCKET}"
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/docker/scripts/container_entry.sh:15]: Do we want this to be GCS dependent? Do we expect customers to create a GCS bucket just to run this? A mount failure will not be fatal, but just checking.

File: sample_workloads/nccltest/docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+# remove NCCL from container
+WORKDIR /third_party/nccl-netsupport
+RUN rm -f build/lib/libnccl*
+WORKDIR /usr/lib/x86_64-linux-gnu/
+RUN rm -f libnccl*
+
+# Install gcsfuse and python3.
+RUN apt-get update \
+  && apt-get install --yes \
+      curl lsb-release cuda-nsight-systems-12-0 python3 python3-pip \
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/docker/Dockerfile:86]: Everything except for curl and lsb-release can be removed; I'm also not sure if we need to have GCS for this.

File: sample_workloads/nccltest/docker/scripts/container_entry.sh [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+: "${WARMUP_ITERS:?Must set WARMUP_ITERS}"
+: "${RUN_ITERS:?Must set RUN_ITERS}"
+: "${N_RUNS:?Must set N_RUNS}"
+
+# Unreserved cores for taskset call. This is a CSV of ranges for cores unused
+# by TCPX.
+: "${UNRESERVED_CORES:?Must set UNRESERVED_CORES}"
+
+# Telemetry.
+: "${GPU_TELEMETRY:?Must set GPU_TELEMETRY}"
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/docker/scripts/container_entry.sh:33]: Remove this.

File: sample_workloads/nccltest/gke/Chart.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v2
+name: nccl-benchmarks
+description: nccl-benchmarks
+type: application
+version: 0.1.0
+appVersion: 0.1.0
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/gke/Chart.yaml:6]: Nit: EOF linebreak.

File: sample_workloads/nccltest/gke/templates/nccl_benchmarks.yaml [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+kind: Service
+metadata:
+  name: "nccl-benchmarks-{{ $.Release.Name }}-{{ $timestamp }}"
+spec:
+  selector:
+    name: "nccl-benchmarks-{{ $.Release.Name }}-{{ $timestamp }}"
+  clusterIP: None
+---
+{{ $node_count := .Values.cluster.nNodes | int }}
+{{ $superblock_count := .Values.cluster.nSuperblocks | int }}
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/gke/templates/nccl_benchmarks.yaml:16]: Concern about the SB portions (lines 16-17, 35-40). This is dependent on the Kubernetes cluster having a `superblock` label indicating the index of the SB, which is baked into our internal GKE cluster creation process.

File: sample_workloads/nccltest/gke/values.yaml [hunk 10]
```diff
@@ -0,0 +1,60 @@
+cluster:
+  nNodes: null  # Must specify on commandline (--set cluster.nNodes=2)
+  sbPlacement: false # Set `true` if running across multiple superblocks
+  nSuperblocks: 1
+  startSuperblock: 1
+  gcsBucket: "pirillo-sct-bucket"
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/gke/values.yaml:6]: Default value not accessible externally (would recommend just removing GCS dependency).

File: sample_workloads/nccltest/gke/values.yaml [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+ncclBenchmarks:
+  image: "us-central1-docker.pkg.dev/supercomputer-testing/pirillo-gcr/nccltest"
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/gke/values.yaml:9]: Default value not accessible externally (TODO: make this an image under us-docker.pkg.dev/gce-ai-infra/...)

File: sample_workloads/nccltest/gke/values.yaml [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  tag: "latest"
+  # A comma-separated list of benchmarks to run.
+  benchmarks: "all_gather_perf,all_reduce_perf,sendrecv_perf"
+  # A comma-separated list of hex masks for the benchmarks.
+  # Must be of the same length as ncclBenchmarks.benchmarks.
+  # Each mask must be less than ncclBenchmarks.gpusPerNode.
+  masks: "0x0,0x0,0x0"
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/gke/values.yaml:16]: For send/recv, make the mask 0x7 to make for a more representative traffic pattern.

File: sample_workloads/nccltest/gke/values.yaml [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  msgSizeEnd: "8G"
+  # Number of GPUs per node. Must be one of 1, 2, 4, 8.
+  gpusPerNode: 8
+  nComms: 1
+  warmupIters: 5
+  runIters: 100
+  nRuns: 5
+
+telemetry:
+  gpu: false
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/gke/values.yaml:27]: Remove this.

File: sample_workloads/nccltest/gke/values.yaml [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    NCCL_MAX_NCHANNELS: 12
+    NCCL_MIN_NCHANNELS: 12
+    NCCL_GPUDIRECTTCPX_SOCKET_IFNAME: "eth1,eth2,eth3,eth4"
+    NCCL_GPUDIRECTTCPX_CTRL_DEV: "eth0"
+    NCCL_GPUDIRECTTCPX_PROGRAM_FLOW_STEERING_WAIT_MICROS: "1000000"
+    NCCL_CROSS_NIC: 0
+    NCCL_ALGO: "Ring"
+    NCCL_PROTO: "Simple"
+    NCCL_NET_GDR_LEVEL: "PIX"
+    NCCL_P2P_PXN_LEVEL: 0
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/gke/values.yaml:60]: Nit: EOF linebreak.

File: sample_workloads/nccltest/docker/scripts/container_entry.sh (hunk 7 above)
Comment (Chris113113) [reply sample_workloads/nccltest/docker/scripts/container_entry.sh:33]: Any specific reason we want to remove the Nsight support? 

We have nsight traces exposed as optional in our other sample workflows.

File: sample_workloads/nccltest/docker/Dockerfile (hunk 6 above)
Comment (Chris113113) [reply sample_workloads/nccltest/docker/Dockerfile:86]: Same as below, if we remove Nsight then I'll cut it, but I think it's still work exposing with a flag.

File: sample_workloads/nccltest/gke/templates/nccl_benchmarks.yaml (hunk 9 above)
Comment (Chris113113) [reply sample_workloads/nccltest/gke/templates/nccl_benchmarks.yaml:16]: Correct, I will highlight this in the readme. We have several different flows right now for cluster creation, that each use a different label or mechanism for superblock assignment.

File: sample_workloads/nccltest/docker/root_scripts/tune_net.sh (hunk 4 above)
Comment (Chris113113) [reply sample_workloads/nccltest/docker/root_scripts/tune_net.sh:50]: Added comment to this script, will also call it out in the readme.

File: sample_workloads/nccltest/docker/scripts/run_nccl_benchmark.sh [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    --mca routed direct --allow-run-as-root -np "$NRANKS" \
+    --hostfile "/scripts/hostfiles${NNODES}/hostfile${GPUS_PER_NODE}" \
+    --timeout $(( 3 * 60 * 60 )) \
+    -x LD_LIBRARY_PATH -x PATH \
+    -x "CUDA_VISIBLE_DEVICES=${CUDA_VISIBLE_DEVICES}" \
+    $NCCL_FLAGS \
+    -x "RUN_LOG_DIR=${RUN_LOG_DIR}" \
+    -x "UNRESERVED_CORES=${UNRESERVED_CORES}" \
+    -x "GPU_TELEMETRY=${GPU_TELEMETRY}" \
+    -x "BENCHMARK=${BENCHMARK}" \
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/docker/scripts/run_nccl_benchmark.sh:56]: Add a -x "NCCL_TESTS_SPLIT_MASK=${MASK}" here to actually use the mask.

File: sample_workloads/nccltest/README.md [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+### Finding Results
+
+The container will log the output of Nvidia's nccl-test binaries for each of the tests that are requested in the parameters.
+
+Each test run is logged separately. If you specify a release that uses multiple tests and multiple runs, the logs will be ordered by Test0Run0...Test0RunN...TestNRunN.
+
+An example test output would look like:
+```
+benchmark: all_reduce_perf, mask: 0x0, run 1/1
+# nThread 1 nGpus 1 minBytes 1048576 maxBytes 8589934592 step: 2(factor) warmup iters: 2 iters: 2 agg iters: 1 validation: 0 graph: 0
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/README.md:53]: Maybe also attach a run with more iterations, to reduce variance? But this is only for demonstration purposes, so I don't think it's a big problem.

File: sample_workloads/nccltest/docker/scripts/container_entry.sh [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+        "$BENCHMARK" != "reduce_scatter_perf" && "$BENCHMARK" != "broadcast_perf" && \
+        "$BENCHMARK" != "reduce_perf"         && "$BENCHMARK" != "sendrecv_perf" && \
+        "$BENCHMARK" != "scatter_perf"        && "$BENCHMARK" != "gather_perf" && \
+        "$BENCHMARK" != "alltoall_perf"       && "$BENCHMARK" != "hypercube_perf" ]]; then
+    echo "${BENCHMARK} is not a legal benchmark, aborting..."
+    exit 2
+  fi
+done
+
+for MASK in ${MASKS[@]}; do
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/docker/scripts/container_entry.sh:53]: We don't necessarily have to have this limit for this public version; up to you whether you want to keep it. If you want to change it, also change line 15 of values.yaml and lines 110 and 111 in README.

File: sample_workloads/nccltest/README.md [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+- For a rail-aligned NCCL operation using only 4 GPUs on a VM, `MASK=0x3`.
+
+*To guarantee symmetry in communication pattern across multiple VMs, we require
+the mask to be less than the number of GPUs used per node.*
+
+Message sizes should be specified using 'G', 'M', 'K', or no suffix for bytes. For example 1G == 1024M == (1024 * 1024)K == (1024 * 1024 * 1024). [Source](https://github.com/NVIDIA/nccl-tests/blob/1292b25553bd0384f2faa2965f9d82b99797a348/src/common.cu#L86C1-L120C2).
+
+##### WARMUP_ITERS, and RUN_ITERS
+
+For each iteration, the benchmark will measure the average latency and bus
```
Comment (anthonyhan2) [inline sample_workloads/nccltest/README.md:117]: Change to:

> For each message size, the benchmark will measure the ...

**Review by anthonyhan2 — APPROVED** [review]

**Review by stevenBorisko — APPROVED** [review]

**Pull request merged**